framework/include/base/MooseFunctor.h

```diff
@@ -549,177 +557,392 @@ Functor<T>::clearCacheData()
+FunctorImpl<T>::dot(const ElemSideQpArg & elem_side_qp, const unsigned int state) const
 {
   return evaluateDot(elem_side_qp, state);
 }
 
 /**
- * Class template for creating constants
+ * A non-templated base class for functors that allow an owner object to hold
+ * different class template instantiations of \p Functor in a single container
+ */
+class FunctorBase
+{
+public:
+  FunctorBase() = default;
+  virtual ~FunctorBase() = default;
+
+  ///@{
+  /**
+   * Virtual methods meant to be used for handling functor evaluation cache clearance
+   */
+  virtual void timestepSetup() = 0;
+  virtual void residualSetup() = 0;
+  virtual void jacobianSetup() = 0;
+  virtual bool wrapsNull() const = 0;
+  virtual std::string returnType() const = 0;
+  ///@}
+};
+
+template <typename T>
+class NullFunctor;
+
+/**
+ * This is a wrapper that forwards calls to the implementation,
+ * which can be switched out at any time without disturbing references to
+ * FunctorImpl. Implementation motivated by https://stackoverflow.com/a/65455485/4493669
  */
 template <typename T>
-class ConstantFunctor : public Functor<T>
+class Functor final : public FunctorImpl<T>, public FunctorBase
 {
 public:
-  using typename Functor<T>::FaceArg;
-  using typename Functor<T>::SingleSidedFaceArg;
-  using typename Functor<T>::ElemFromFaceArg;
-  using typename Functor<T>::ElemQpArg;
-  using typename Functor<T>::ElemSideQpArg;
-  using typename Functor<T>::FunctorType;
-  using typename Functor<T>::FunctorReturnType;
-  using typename Functor<T>::ValueType;
-  using typename Functor<T>::GradientType;
-  using typename Functor<T>::DotType;
+  using typename Moose::FunctorImpl<T>::ValueType;
+  using typename Moose::FunctorImpl<T>::GradientType;
+  using typename Moose::FunctorImpl<T>::DotType;
 
-  ConstantFunctor(const ValueType & value) : _value(value) {}
-  ConstantFunctor(ValueType && value) : _value(value) {}
+  /**
+   * Construct wrapper from wrapped object
+   */
+  Functor(const FunctorImpl<T> & wrapped) : FunctorBase(), _wrapped(&wrapped) {}
+  Functor(FunctorImpl<T> &&) = delete;
 
-private:
-  ValueType evaluate(const libMesh::Elem * const &, unsigned int) const override final
+  Functor(std::unique_ptr<FunctorImpl<T>> && wrapped)
+    : FunctorBase(), _owned(std::move(wrapped)), _wrapped(_owned.get())
+  {
+  }
+
+  /**
+   * Assign our wrapped object to be something new and release our previously wrapped object
+   */
+  void assign(const FunctorImpl<T> & wrapped)
   {
-    return _value;
+    _owned.reset();
+    _wrapped = &wrapped;
   }
-  ValueType evaluate(const ElemFromFaceArg &, unsigned int) const override final { return _value; }
-  ValueType evaluate(const FaceArg &, unsigned int) const override final { return _value; }
-  ValueType evaluate(const SingleSidedFaceArg &, unsigned int) const override final
+
+  void assign(std::unique_ptr<FunctorImpl<T>> && wrapped)
   {
-    return _value;
+    _owned.reset();
+    _owned = std::move(wrapped);
```

Comment:
```suggestion
    _owned.reset(wrapped);
```
this works?

Reply:
I get compile errors with gcc if I try this

Reply:
But I can remove the `reset` call and just do the assignment as that destructs the previously owned object on its own